pg-eng-Aim-Sliding-on-Controller-LV-2226

nrice25 opened the pull request:
[Jira Ticket](https://bradleycapstone.atlassian.net/browse/LV-2226?atlOrigin=eyJpIjoiMmU0M2ZjYmQ0MjI3NDUzYjk0ODJkZjE0MDI1MTdlMjMiLCJwIjoiaiJ9)

Estimated review time: 15 minutes

There is now an event that fires whenever the controller and keyboard swap 
There should now be adjustable variables for horizontal aim speed

Thanks for the help Ryan

nrice25 committed:
Aim-Sliding-on-Controller

There is now an event that fires whenever the controller and keyboard swap
There should now be adjustable variables for horizontal aim speed

nrice25 committed:
Merge branch 'development-post-grad' into pg-eng-Aim-Sliding-on-Controller-LV-2226

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+using UnityEngine.InputSystem;
+using UnityEngine.InputSystem.Users;
+
+public class ControlsManager : MainUniversalManagerFramework
```
XML needed here

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+    public static ControlsManager Instance;
+    private readonly UnityEvent<InputDevice> _onChangeControls = new();
+    private InputDevice _currentDevice;
+    private bool canUseControlSwap = true;
```
Should start with an underscore

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -601,6 +610,22 @@ public void RestartAutoCameraMovement()
+        }
+        else if (device == Mouse.current || device == Keyboard.current)
+        {
+            Debug.Log("GHOSTS");
```
I love these debugs are just completely random but they gotta go.

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -601,6 +610,22 @@ public void RestartAutoCameraMovement()
+        {
+            _virtualCamera.GetCinemachineComponent<CinemachinePOV>().m_HorizontalAxis.m_AccelTime =
+                _horizontalAccelTimeMK;
+            Debug.Log("HOWS");
```
Same with this debug

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -601,6 +610,22 @@ public void RestartAutoCameraMovement()
         }
     }
 
+    private void ChangeCameraAccelTime(InputDevice device)
```
XML comment

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+        InputUser.onChange += ControlSwap;
+    }
+
+    private void ControlSwap(InputUser user, InputUserChange userChange, InputDevice device)
```
XML comment

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+    /// <summary>
+    /// This waits for the end of the frame before allowing the control swap to activate again
+    /// </summary>
+    private IEnumerator WaitForThing()
```
Maybe should have a different name than WaitForThing

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+    #endregion
+}
+
+internal class PlayerInputDevice : InputDevice
```
Could probably use a class header, though I am confused as to what this class is used for?

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -223,7 +230,9 @@ private IEnumerator MoveCamera()
 
+            //yield return new WaitForEndOfFrameUnit();
             yield return null;
+            //yield return new WaitForFixedUpdate();
```
Could delete this commented line

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -223,7 +230,9 @@ private IEnumerator MoveCamera()
             AdjustHarpoonRotation();
             BoatSway();
 
+            //yield return new WaitForEndOfFrameUnit();
```
Also could delete this commented line

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -601,6 +610,22 @@ public void RestartAutoCameraMovement()
+    {
+        if (device == Gamepad.current)
+        {
+            _virtualCamera.GetCinemachineComponent<CinemachinePOV>().m_HorizontalAxis.m_AccelTime =
```
We could maybe store the CinemachinePOV as a variable, but not mandatory

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -124,6 +124,13 @@ public class PlayerCameraController : MonoBehaviour
+    [SerializeField] private float _horizontalAccelTimeController = .5f;
+    
+    [SerializeField] private float _verticalAccelTimeMK = 1f;
+    [SerializeField] private float _verticalAccelTimeController = .3f;
```
These 2 variables aren't being used, though I recognize they probably should be used in ChangeCameraAccelTime()

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+
+    private void ControlSwap(InputUser user, InputUserChange userChange, InputDevice device)
+    {
+        if (device == null || _currentDevice == device || !canUseControlSwap)
```
Could maybe add a comment to explain the thought process. I imagine anyone except you and me would be confused by this line

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -0,0 +1,77 @@
+    // Start is called before the first frame update
+    void Start()
+    {
+        InputUser.onChange += ControlSwap;
```
This is never unsubscribed. Also we could probably use the SubscribeToEvents and UnsubscribeToEvents functions from MainUniversalManagerFramework.
IMPORTANT: I just checked to see if I was crazy, I tried moving this subscription over to the SubscribeToEvents function and the errors I mentioned in Discord don't occur.

RyanASwanson commented on Assets/Scripts/PlayerScripts/PlayerCameraController.cs:
```diff
@@ -124,6 +124,13 @@ public class PlayerCameraController : MonoBehaviour
+    [SerializeField] private float _horizontalAccelTimeController = .5f;
+    
+    [SerializeField] private float _verticalAccelTimeMK = 1f;
+    [SerializeField] private float _verticalAccelTimeController = .3f;
```
Also Cinemachine has a value for deceleration too, might be nice to add variables to modify that as well. Not required, but recommended

nrice25 replied:
I would say that should be a task for the next sprint; or the one after that. We need to ask the designers what they want from this new system.

RyanASwanson reviewed (changes requested):
Primary reason for request changes is due to the errors in console, in one of my comments I explained the cause. Also the code isn't actually modifying the vertical acceleration at all. Other those its mostly just code standards comments. I left a good few comments but most of them should take 5 seconds to fix

nrice25 committed:
Merge branch 'development-post-grad' into pg-eng-Aim-Sliding-on-Controller-LV-2226

nrice25 committed:
Code cleanup

Vertical movement is now changed alongside horizontal
The events should be properly subscribed and unsubscribed

nrice25 committed:
Added more comments

RyanASwanson commented on Assets/Scripts/Managers/UniversalManagers/MainUniversalManagers/ControlsManager.cs:
```diff
@@ -67,11 +72,16 @@ public override void SetUpInstance()
+        InputUser.onChange += ControlSwap;
+    }
+
+    protected override void UnsubscribeToEvents()
```
Even when overriding it we generally add an XML regardless, couldn't hurt

RyanASwanson reviewed (approved):
This looks good to me. I left a single comment about having an XML comment. I still stand by my original request for deceleration variables but I won't require it. I just feel that it can't hurt to have, as we can just set the variables to the default deceleration for now and design can adjust it if they feel it appropriate.

nrice25 committed:
Merge branch 'development-post-grad' into pg-eng-Aim-Sliding-on-Controller-LV-2226

nrice25 committed:
Comments and pseudo deccel addition

Xml comments for subscribe and unsubscribe
Added in the variables for deceleration and commented it out in the function
